[1단계 - 출석 구현] 새로이(강철원) 미션 제출합니다.

Ryan-Dia opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [링크](https://prolog.techcourse.co.kr/studylogs/3917)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->

  


❗️ 실행 테스트 하실 때 controller 맨 위에 있는 변수인 fixDateTime에서 25를 16으로 수정해주세요. 
`LocalDateTime fixDateTime = LocalDateTime.of(2024, Constants.MONTH, 25, 0, 0, 0, 0);`


<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

안녕하세요 제로! 만나서 반갑습니다.  

뭔가 닉네임이 익숙해서 찾아봤는데 알고보니 `[10분 테코톡] 라라, 제로의 데이터베이스 인덱스` 해당 영상에서 뵈었던 분이었습니다. 
실제로 우테코 들어오기전에 DB공부할 때 몇번을 봤던 영상이었는데 도움이 많이 되었습니다. 
그러던 중 궁금증이 생겨서 레벨 1은 어떻게 보내셨는지 보았는데 아래와 같은 글을 발견하고 공감이 많이 되었습니다.

>페어와 의견이 다르면 구현을 멈추고 토론을 진행했다. 컴퓨터공학과를 전공하였지만 처음 있는 경험이다.
>
>생각하는 방법을 설명하고 한 번 더 내용을 상기한다.
>무엇이 더 좋은 접근법인지 고민한다.
>처음 의견이 갈렸을 때는 ‘크루의 기분이 나쁘지 않을까’라는 걱정이 먼저 들었다. 걱정도 잠시일 뿐이다. 페어에게 의견을 내놓았을 때 의견을 
>존중한다는 느낌을 받았다. 자연스레 자신의 생각을 더 공유할 수 있게 되었다.
>
>정신없이 토론하고 구현하다 보니 페어프로그래밍을 시작한 화요일이 금요일로 바뀌었다. 서로의 생각이 다르기에 더 좋은 코드가 나온 것이다. >배움 뿐 아니라 좋은 에너지도 받았다. 나도 다른 크루들에게 좋은 에너지를 전파하는 사람이 되자고 다짐했다.
>
>나는 성장하고 있다. 스스로도 성장했다고 느낀다. 하지만, 아직 부족한 점이 매우 많다. 남은 기간동안에도 더 배움의 자세로 우테코에 임하자. >11월까지 모두 화이팅했으면 좋겠다!

저도 눈떠보니 어느새 화요일에서 목요일이 되어있었습니다. 
이번에 처음 TDD로 하다보다 속도도 굉장히 느리고 구현할 때마다 "이건 나중에 리팩터링 하면 될 것 같아요"를 구현하면서 계속 서로 외쳤는데 결국에는 리팩터링을 거의 하지 못하고 끝나게 되어서 아쉽고 리뷰어분에게 죄송스럽기도 합니다. 
그래서 아쉬운 점들은 아래 글로 정리해보았습니다. 

최종 코테 5시간보다 더 퀄리티가 떨어지는 것 같아 보시기 힘드실 것 같아서 마음이 많이 무겁습니다.
그럼에도 불구하고 그래도 잘 부탁드립니다. 😊

## 😰 아쉬운 점

### 0. 처음 TDD 진행해서 리팩터링을 거의 못한 점
처음으로 TDD로 페어프로그래밍을 해보았습니다. 
처음에는 어려웠지만 아주 조금씩 감을 잡아가면서 속도가 붙었지만 결국에는 리팩터링을 할 시간까지 확보를 못해서 아쉬웠습니다.

### 1. 휴일 관리 객체를 만들지 못한 점 
메서드 내에서 관리하기 보다는 휴일(주말, 공휴일)을 전체적으로 관리하는 객체 만들어 관리를 해야하는 필요성을 느낌

### 2. Week enum 클래스의 책임 분리
현재 Week enum에서는 크게 두가지의 기능을 처리하고 있습니다.

1. 요일에 따른 출석 시간 관리
2. 요일 한글화

시간 관계상, 요일 한글화 기능 구현을 가장 비슷한 역할을 가진 Week 추가하였습니다.

### 3. Controller의 비대화
Controller에서 모든 도메인의 흐름을 담당하고 있는데
이 때문에 종종 비지니스의 로직이 섞여 가독성을 저해시키고, 복잡성을 증가시킴

### 4. 시간 자료형에 대한 포멧팅
미션 내에 LocalDate 와 LocalDateTime에 대한 입출력 로직이 많은데
이를 util로 관리하면 어땠을까 하는 아쉬움이 남았습니다.

## 💡리뷰 받고 싶은 포인트들

### 0️⃣ 크루 와 출석의 위치

<img width="358" alt="자료1" src="https://github.com/user-attachments/assets/6899f673-6a30-48d0-b53a-31906489a506" />


현재 작성된 코드로는 크루 내에 출석s 가 존재하는 구조입니다.

하지만 아래와 같이 구조를 수정하는 방식도 생각을 해봤습니다.
![자료2](https://github.com/user-attachments/assets/5458c6e9-f942-48e4-bd90-f877e97f352c)


장점: depth를 깊지 않게 관리할 수 있음

but, depth가 그렇게 큰 이유인 지 모르겠어서 depth 만이 이유라면 수정의 필요성이 클까 싶기도 합니다.

### 1️⃣ InputView 메서드 내에서 String 반환 (의존성 제거)

페어와 InputView 메서드 반환 타입에 대해 생각이 달랐습니다.

아래 내용은 InputView 중 `readNickName` 메서드를 기준으로 작성했습니다.

1. String 자료형 반환 (현 코드 로직)

```java
public static String readNickName() {
        System.out.println("닉네임을 입력해 주세요.");
        String input = scanner.nextLine();
        validateInput(input);
        return input;
    }
```

- 장점: View에 도메인 의존성 분리
- 단점: controller 에 도메인 생성자까지 나열

1. 도메인 반환 (의견이 나왔던 코드 로직)

```java
public static Nickname readNickName() {
        System.out.println("닉네임을 입력해 주세요.");
        String input = scanner.nextLine();
        validateInput(input);
        return new Nickname(input);
    }
```

- 장점: controller에 도메인 생성자까지 나열하지 않아도 됨
- 단점: View에 도메인 의존성 주입

### 2️⃣ 같은 로직을 처리하는 메서드 관리

미션에서 출력 안내 문구만 다를 뿐 같은 로직을 처리하는 메서드가 두 개 있었습니다.

**기능 1**

```
닉네임을 입력해 주세요.
이든
등교 시간을 입력해 주세요.
09:59
```

**기능 2**

```
출석을 수정하려는 크루의 닉네임을 입력해 주세요.
빙티
수정하려는 날짜(일)를 입력해 주세요.
3
언제로 변경하겠습니까?
09:58
```

여기서

```java
- 크루의 닉네임을 받는 기능
- 등교 시간을 받는 기능
```

이렇게가 출력 안내 문구만 다를 뿐 같은 기능을 하고 있습니다.

하지만 코드 내에서는 이를 효율적으로 처리하지 못했습니다.

**크루 닉네임을 받는 로직들**

```java
public static String readNickName() {
        System.out.println("닉네임을 입력해 주세요.");
        String input = scanner.nextLine();
        validateInput(input);
        return input;
    }

    public static String readUpdateNickName() {
        System.out.println("출석을 수정하려는 크루의 닉네임을 입력해 주세요.");
        String input = scanner.nextLine();
        validateInput(input);
        return input;
    }
```

**등교 시간을 받는 로직들**

```java
public static String readDateTime() {
        System.out.println("등교 시간을 입력해 주세요.");
        String input = scanner.nextLine();
        validateInput(input);
        return input;
    }

    public static String readUpdateDateTime() {
        System.out.println("언제로 변경하겠습니까?");
        String input = scanner.nextLine();
        validateInput(input);
        return input;
    }
```

그렇다면 매개 인자로 안내 문구를 받아 공통의 메서드로 수정하는 것을 생각해봤는데 이 방식이 더 나을까요??

- 매개 인자로 안내 문구를 받는 로직

```java
public static String readDateTime(String readTimeGuide) {
        System.out.println(readTimeGuide);
        String input = scanner.nextLine();
        validateInput(input);
        return input;
    }
```

### 3️⃣ 정렬 위치 view?

기능 3이 한 크루의 출석부를 날짜 오름차순으로 정렬해야 했습니다.

```java
public static void printCrewAttendances(Crew crew) {
        final Nickname nickname = crew.getNickname();
        final Attendances attendances = crew.getAttendances();
        attendances.sort();

        // ...
    }
```

여기서 고민했던 부분은 정렬을

1. LinkedList로 관리하여 처리할 건지
2. 도메인 내에서 sort를 진행할 건지
3. view 내에서 sort를 진행할 건지

네오의 강의 내에서 2번 vs 3번 에 대한 의견 교류 시간이 있었습니다.

이 때 네오는 둘 다 사용이 가능하나 각자의, 책임에 대한 이유에 따라 나뉠 수 있는 부분이라고 말해주셨습니다.

### 4️⃣ 출석 가능한 일자 뽑아오는 로직의 효율성

```java
등교하지 않아 출석 기록이 없는 날은 결석으로 간주한다.
```

출석을 해야 하지만 등교를 하지 않은 날을 추출하기 위해 다음과 같은 로직을 구현하였습니다.

**📌 CrewGenerator class 내 코드**

```java
public static List<Integer> getValidDates(LocalDate localDate) {
        final int today = localDate.getDayOfMonth();

        List<Integer> allDays = IntStream.range(1, today).boxed().collect(Collectors.toList());
        allDays.removeAll(getExcludeNotAttendanceDays());

        return allDays;
    }

    public static List<Integer> getExcludeNotAttendanceDays() {
        List<Integer> excludeNotAttendanceDays = new ArrayList<>();
        excludeNotAttendanceDays.addAll(getWeekendDays());
        excludeNotAttendanceDays.addAll(Constants.HOLIDAYS);

        return excludeNotAttendanceDays;
    }

    private static Collection<Integer> getWeekendDays() {
        return IntStream.rangeClosed(1, Constants.LENGTH_OF_MONTH)
                .filter(CrewGenerator::excludeNotAttendanceDays)
                .boxed()
                .toList();
    }

    public static boolean excludeNotAttendanceDays(int day) {
        DayOfWeek dayOfWeek = LocalDate.of(2024, Constants.MONTH, day).getDayOfWeek();
        return dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY;
    }
```

위의 코드 로직은 등교 해야 하는 날을 추출하는 모든 로직입니다.

로직 순서는 다음과 같습니다.

1. 1부터 31까지의 숫자가 들어있는 list 생성 (:allDays)
2. 출석하지 않는 날이 들어있는 list 생성(: excludeNotAttendanceDays)
    
    ```java
    public static List<Integer> getExcludeNotAttendanceDays() {
            List<Integer> excludeNotAttendanceDays = new ArrayList<>();
            excludeNotAttendanceDays.addAll(getWeekendDays());
            excludeNotAttendanceDays.addAll(Constants.HOLIDAYS);
    
            return excludeNotAttendanceDays;
        }
    ```
    

위 로직에서 연산 횟수를 고려하면, allDays는 최대 31번, getWeekendDays 또한 최대 31번 수행되므로 총 **62번의 연산이 발생**한다고 예상했습니다.

그러나 getWeekendDays 로직을 다음과 같이 수정하면 연산 횟수를 줄일 수 있을 것이라고 고민했습니다.

1.	**첫 번째 토요일과 일요일을 찾기**

2.	**각 요일에 +7씩 더하면서 리스트에 추가하기**

혹시 리뷰어님은 해당 로직을 작성해야한다면 어떤 부분을 포인트로 잡고 구현하실 생각인지 고견을 듣고 싶습니다.

### 5️⃣ class 이름과 같은 필드 이름

다음 코드는 `NickName class` 에 일부분을 가져왔습니다.

```java
public class Nickname implements Comparable<Nickname> {

    private final String nickname;

    //...
```

다음 코드는 `Crews class` 에서 일부분을 가져왔습니다.

```java
public class Crews {

    private List<Crew> crews;
    
    //...
}
```

이렇게 클래스의 이름과 필드 이름이 동일할 때가 종종 발생했습니다.

페어와 이 부분에 대한 필드명 고민을 진행했었는데요,

`value`, `name` 는 사용할 수 없겠다라는 결론을 내리게 되었습니다.

- `value`
    - 필드명에서 정확한 의미를 담지 못함
- `name`
    - Enum에서 사용되는 예약어임

이럴 때는 어떤 명명을 주로 사용하시는 지 궁금합니다.

### 6️⃣ Test 코드, DisplayName 애노테이션 사용 유무

페어와 의견이 달랐던 부분인데요,

현재 테스트 코드에서는 DisplayName 애노테이션을 사용하지 않고 한글 메서드명을 사용하고 있습니다.

```java
@Test
    void 출석이_정상적으로_실행() {
        // given
        final LocalDateTime localDateTime = LocalDateTime.of(2024, 12, 2, 0, 0, 0);

        // when
        // then
        assertThatCode(() -> new Attendance(localDateTime))
                .doesNotThrowAnyException();
    }

```

테스트 작성 시, 

DisplayName 애노테이션 을 썼을 때의 장단점에 대해 토론 해본 결과가 아래와 같이 도출되었습니다.

- DisplayName 애노테이션 을 썼을 때
    - 장점
        - 한글 컴파일을 지원하지 않는 프레임워크에서도 동작 가능
            
            ⇒ 예기치 못한 에러 가능성 없엠
            
    - 단점
        - 메서드명을 영어로도 한 번 더 써줘야 함

현업에서는 어떠한 방식을 주로 쓰는 지, 주로 쓰지 않더라도 더 고민해볼 만한 근거가 있으실까요?

### 7️⃣ 출력 시, StringFormat vs “\n” 매번 작성

현재 출력 시, 기본적으로 `System.*out*.println` 를 사용하고 있습니다.

`String.format` 으로 문자열 형식을 지정해주고 있는데

```java
System.out.println(String.format("%s 대상자입니다.", punishment.getPunishmentName()));
```

다른 방식으로는 

```java
System.out.printf("%s 대상자입니다.\n", punishment.getPunishmentName());
```

혹시 현업에서는 이런 비슷한 상황이 있을 때 어떻게 작성하시는 지 궁금합니다. 
저희 둘의 방식말고 또 다른 좋은 방법이 있는지도 궁금합니다.

### 8️⃣ 의존성 depth 때문에 같은 이름의 메서드 계속 생성

```java
public class Crew implements Comparable<Crew> {

    Nickname nickname;
    Attendances attendances;
    AttendanceCounter attendanceCounter;

		// ...

    public Nickname getNickname() {
        return nickname;
    }
    
    // ...
}
```

```java
public class Nickname implements Comparable<Nickname> {

    private final String nickname;

    // ...

    public String getNickname() {
        return nickname;
    }
    
    // ...
}
```

`Crew` 내에 `NickName class`를 필드로 가지고 있어서

`Crew`의 닉네임을 가져오고 싶을 때는 

`crew.getNickname().getNickname()` 이런식으로 가져오게 됩니다.

여기서 궁금한 점은 크게 2가지가 있습니다.

1. `Crew`에서 `NickName`에게 가야 만 받아올 수 있는 지,
    - 제 생각으로는 받아올 수 밖에 없지만,, 혹시 가능한 방법이 있는 지 궁금합니다. crew 내에서 그냥 한 번에 처리 할 수 있는 지,,
2. 만약 받아올 수 밖에 없다면 `crew`에서의 getNickname() 과 Nickname 내에서의 getNickname() 명명을 어떻게 수정하면 좋을 지
    - Nickname에서는 getNickName이 getter의 기본 명명 규칙과 같으므로 수정할 이유가 없다고 생각되고
    - 중복되어 바꿔야 한다면 Crew의 getNickname()의 이름을 수정해야 된다고 생각하는데
        - getNicknameClass() 도 어색한 것 같아 더 좋은 이름을 찾고 싶습니다.
        

저의 생각은 `Nickname`  클래스에서 nickname 필드 이름을 displayNickname으로 바꿔서 `Crew` 객체에서 `getDisplayNickname()` 이란 메서드를 만들어서 가져올려고 했습니다.

### 9️⃣ 명령어 입력 controller if 처리

지금 컨트롤러의 if분기는 아래와 같습니다.  

```java
if (command.equals(Command.QUIT)) {
  break;
}
if (command.equals(Command.CHECK_ATTENDEES)) {
    processCheckAttendees(crews, fixDateTime);
    continue;
}
if (command.equals(Command.EDIT_ATTENDANCE)) {
    processEditAttendance(crews);
    continue;
}
if (command.equals(Command.CHECK_THE_ATTENDANCE_RECORD_BY_CREW)) {
    processAttendanceRecordByCrew(crews);
    continue;
}
if (command.equals(Command.CONFIRMATION_OF_THOSE_AT_RISK_OF_EXPULSION)) {
    OutputView.printAllExpulsion(crews);
}
```

이렇게 만들면 명령어가 늘어날 수록 if문이 늘어나야합니다.  

여기서 어떻게 하면 변경에 대한 유연성을 높이면서 효율적으로 짤 수 있는 방법에 대한 힌트를 주시면 감사합니다.  

저희도 계속 생각해보겠습니다.

Ryan-Dia committed:
docs: 기능 목록 추가

Co-authored-by: jin
<2jin2.1031@gmail.com>

Ryan-Dia committed:
chore: 출석 csv파일 추가

Co-authored-by: jin
<2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 출석 상태 enum 추가

Ryan-Dia committed:
feat: 요일별 출석 시간 관리 enum 추가

Ryan-Dia committed:
feat: 출석 기능 구현

Co-authored-by: jin
<2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 닉네임 객체화

Co-authored-by: jin
<2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 요일별 출근 시간 enum 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 기능 명령 enum 관리 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
fix: 출근 시간 판별 메서드에 경계선 값 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 출석 개수 카운팅 메서드 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 주말 출석 제외 기능 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 크루 일급컬렉션 생성

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: csv리더기 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 처벌 관리 enum 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat : 크루 생성 로직 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 출석 개수 관리 객체 생성

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 데이터로 크루그룹 생성 기능 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 출석dto 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 출력을 위한 getter 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 요일 한글화 기능 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 크루 정렬 기능 추가(처벌 기준에 따른)

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 운영시간 처리 기능 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
test: 유효한 테스트 조건으로 변경

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: 에러 출력 메세지 포멧 수정

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 크루 출석 관리 기능들 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 닉네임 equals 및 hashCode 재정의

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 출석 관리 기능들 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 출석 관리 기능들 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 존재하지 않는 출석일 처리

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: 크루 생성기 폴더 위치 수정

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 출석 controller 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: view 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
chore: application 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: 사용하지 않는 메서드 제거

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: 명령어 출력에서 Command enum 활용

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: 공통 상수로 교체

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: controller 메서드 분리

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: compareTo 메서드 재정의

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
feat: 정렬기준 추가(이름순)

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
fix: 출력 형식수정

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
fix: 출력 형식 수정

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
style: 메서드 순서 내림차순으로 변경

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
fix: 고정날짜 16로 변경

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: 사용하지 않는 메서드 삭제

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
style: 불필요한 공백 제거 및 필요한 공백 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
fix: 출력 형식 수정

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: 불필요한 코드 삭제

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: final 키워드 추가

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: 에러메시지 형식 수정

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
style: 매직넘버 상수화

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: static import 방식 수정

Co-authored-by: 2Jin1031 <2jin2.1031@gmail.com>

Ryan-Dia committed:
refactor: processCheckAttendees 내부 로직 메서드로 분리

Ryan-Dia committed:
fix: 기능 선택시 해당 기능 무한반복이 아닌 전체 로직 무한 반복으로 변경

Ryan-Dia committed:
refactor: csvPath 상수객체로 이동

Ryan-Dia committed:
refactor: final 키워드 추가

Ryan-Dia committed:
style: 변수명 변경(MONTH->FIXED_MONTH)

Ryan-Dia committed:
refactor: processEditAttendance 내부 로직 메서드 분리

Ryan-Dia committed:
refactor: 내부 메서드를 사용하도록 수정

Ryan-Dia committed:
style: 매직넘버 상수로 교체

Ryan-Dia commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,92 @@
+import view.InputView;
+import view.OutputView;
+
+public class AttendanceController {
```
controller부터 보기시 너무 힘드실 것 같아 아래와 같이 리팩터링을 해보았습니다. 

- 불필요한 static 키워드 제거
- final을 통일감 있게 사용하기
- 매직넘버 상수화
- 중복로직을 메서드화 시켜서 메서드를 호출하기
- 하나의 기능을 반복하는 게 아닌 전체 기능을 반복하도록 수정
- controller에서 불필요한 정적 변수 상수객체로 이동

```java
public class AttendanceController {

    public void start() {
        while (true) {
            final LocalDateTime fixDateTime = LocalDateTime.of(Constants.FIXED_YEAR, Constants.FIXED_MONTH,
                    Constants.FIXED_DAY, 0, 0,
                    0, 0);
            final String input = InputView.readCommand(fixDateTime);
            final Command command = Command.findByCommandNumber(input);
            final Crews crews = CrewGenerator.generate(CsvReader.readFile(Constants.CSV_PATH),
                    fixDateTime.toLocalDate());

            if (command.equals(Command.QUIT)) {
                break;
            }
            if (command.equals(Command.CHECK_ATTENDEES)) {
                processCheckAttendees(crews, fixDateTime);
                continue;
            }
            if (command.equals(Command.EDIT_ATTENDANCE)) {
                processEditAttendance(crews);
                continue;
            }
            if (command.equals(Command.CHECK_THE_ATTENDANCE_RECORD_BY_CREW)) {
                processAttendanceRecordByCrew(crews);
                continue;
            }
            if (command.equals(Command.CONFIRMATION_OF_THOSE_AT_RISK_OF_EXPULSION)) {
                OutputView.printAllExpulsion(crews);
            }
        }
    }

    private void processCheckAttendees(final Crews crews, final LocalDateTime fixDateTime) {
        final Nickname nickname = readNickname();
        final Crew crew = crews.findByNickname(nickname);
        final LocalDateTime attendedDateTime = LocalDateTime.of(fixDateTime.toLocalDate(), readLocalTime());

        if (crew.isAttended(attendedDateTime)) {
            throw new CustomIllegalArgumentException("이미 출석했습니다. 다음에는 수정기능을 이용해주세요.");
        }
        final Attendance attendance = attend(crew, attendedDateTime);

        OutputView.printAttendance(attendance);
    }

    private Nickname readNickname() {
        final String inputNickName = InputView.readNickName();
        return new Nickname(inputNickName);
    }

    private LocalTime readLocalTime() {
        final String inputTime = InputView.readDateTime();
        final DateTimeFormatter formatter = DateTimeFormatter.ofPattern(Constants.TIME_FORMAT);
        return LocalTime.parse(inputTime, formatter);
    }

    private Attendance attend(Crew crew, LocalDateTime attendedDateTime) {
        final Attendance attendance = new Attendance(attendedDateTime);
        crew.add(attendance);
        return attendance;
    }


    private void processEditAttendance(final Crews crews) {
        final Nickname nickname = readNicknameForEditAttendance();
        final Crew crew = crews.findByNickname(nickname);
        final LocalDateTime desiredUpdateDateTime = readUpdateDateTime();
        final Attendance oldAttendance = crew.getAttendance(desiredUpdateDateTime);
        final Attendance newAttendance = new Attendance(desiredUpdateDateTime);

        crew.updateAttendance(oldAttendance, newAttendance);
        OutputView.printUpdateAttendance(oldAttendance, newAttendance);
    }

    private Nickname readNicknameForEditAttendance() {
        final String inputNickName = InputView.readUpdateNickName();
        return new Nickname(inputNickName);
    }

    private LocalDateTime readUpdateDateTime() {
        final String updateDate = InputView.readUpdateDate();
        final LocalTime desiredUpdateTime = readDesiredUpdateTime();
        final LocalDate fixedLocalDate = LocalDate.of(Constants.FIXED_YEAR, Constants.FIXED_MONTH,
                Integer.parseInt(updateDate));
        return LocalDateTime.of(fixedLocalDate, desiredUpdateTime);
    }

    private LocalTime readDesiredUpdateTime() {
        final String inputDateTime = InputView.readUpdateDateTime();
        try {
            return LocalTime.parse(inputDateTime);
        } catch (DateTimeParseException e) {
            throw new CustomIllegalArgumentException("올바른 시간 형식이 아닙니다. 얘를들어 10:22 이런 형태로 작성해주세요.");
        }
    }

    private void processAttendanceRecordByCrew(final Crews crews) {
        Nickname nickname = readNickname();
        Crew crew = crews.findByNickname(nickname);

        OutputView.printCrewAttendances(crew);
    }
}
```

```java
public final class Constants {

    private Constants() {
    }

    public static final int FIXED_YEAR = 2024;
    public static final int FIXED_MONTH = 12;
    public static final int FIXED_DAY = 16;
    public static final int LENGTH_OF_MONTH = 31;
    public static final List<Integer> HOLIDAYS = List.of(25);
    // 00:00은 고정된 운영시간에서 불가능한 출석시간이기에 이 시간을 결석 대체 시간으로 사용
    public static final LocalTime ABSENCE_TIME = LocalTime.of(0, 0);
    public static final String TIME_FORMAT = "HH:mm";
    public static final String CSV_PATH = "src/main/resources/attendances.csv";
}

```

kang-hyungu replied:
리팩터링한 코드는 리뷰 참고해서 반영해보시면 좋겠네요.

kang-hyungu replied:
그리고 Constants 클래스는 제거하세요.

asebn1 replied:
`readNickname`과 `readNicknameForEditAttendance`는 동일한 역할을 하고있네요

Ryan-Dia reviewed (commented):
제일 처음 보는 controller부터 보기 힘드실 것 같아서 리팩터링을 조금 해보았습니다.

Ryan-Dia committed:
fix: 출석 수정 로직에서 발생하는 포멧팅에러타입 변경

kang-hyungu commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,92 @@
+import view.InputView;
+import view.OutputView;
+
+public class AttendanceController {
```
컨트롤러라고 이름 붙인 이유가 무엇인가요?

kang-hyungu commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,92 @@
+                fixDateTime.toLocalDate());
+
+        while (true) {
+            if (command.equals(Command.QUIT)) {
```
Command에 새로운 기능이 추가될 때마다 if절이 점점 늘어나겠네요
더 간결하게 관리할 방법 없을까요?

kang-hyungu commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,92 @@
+    private static String CSV_PATH = "src/main/resources/attendances.csv";
+
+    public void start() {
+        LocalDateTime fixDateTime = LocalDateTime.of(2024, Constants.MONTH, 25, 0, 0, 0, 0);
```
왜 Constants.MONTH가 12월인가요?
상수명만 봐서는 12월인지 알 방법이 없네요.

kang-hyungu commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,92 @@
+    private static String CSV_PATH = "src/main/resources/attendances.csv";
+
+    public void start() {
+        LocalDateTime fixDateTime = LocalDateTime.of(2024, Constants.MONTH, 25, 0, 0, 0, 0);
```
왜 변수명이 fixDateTime 인가요? 뭘 고정하는건가요?

asebn1 replied:
추가로 `fixDateTime`를 상수화도 고려하면 좋을 것 같아요~

Ryan-Dia replied:
감사합니다.
fixDateTime은 SYSTEM_DATE_TIME으로 상수화를 진행했습니다.
또한 이 부분만큼은 변수로 의미를 모두 표현하기 힘들다고 판단해서 프로젝트 특성상 주석을 추가해주었습니다.

kang-hyungu commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,53 @@
+    AttendanceStatus attendanceStatus;
+
+    public static Attendance of(final String inputTime) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DATE_TIME_FORMAT);
```
요건 private static final로 관리해도 괜찮겠네요

Ryan-Dia replied:
감사합니다. 
TDD할 때 필드까지만 만들고 온 집중이 테스트에 가다보니 처음으로 접근제어자와 final을 곳곳에서 빼먹었습니다. 
현재 모든 부분 수정했습니다. 😁

kang-hyungu commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,53 @@
+            LocalDateTime dateTime = LocalDateTime.parse(inputTime, formatter);
+            return new Attendance(dateTime);
+        } catch (DateTimeParseException e) {
+            throw new CustomIllegalArgumentException("올바른 형식이 아닙니다.");
```
예외 클래스명을 Custom을 붙이지 말고 적절한 이름으로 바꿔보세요
올바른 형식이 아니라면 어떤 형식이 올바른 것인가요?

Ryan-Dia replied:
이름에 대해서는 조금 더 생각해 보겠습니다. 

예외메시지는 올바른 형식의 예제를 추가적으로 표시하도록 수정하였습니다.

```
"올바른 형식이 아닙니다. 이와 같은 형식으로 시간을 적어주세요 ex) 10:00"
```

kang-hyungu commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,53 @@
+        final int day = localDateTime.getDayOfMonth();
+        if (Constants.HOLIDAYS.contains(day)) {
+            throw new CustomIllegalArgumentException(
+                    String.format("%d월 %d일 %s은 등교일이 아닙니다.", Constants.MONTH, localDateTime.getDayOfMonth(),
```
String.format(...) 이 부분을 변수로 추출하면 줄바꿈을 안 해도 되겠네요.

Ryan-Dia replied:
감사합니다. 해당부분 변수로 추출했습니다.!

kang-hyungu commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,53 @@
+        }
+    }
+
+    public int getDate() {
```
반환하는 값은 DayOfMonth인데 왜 메서드명은 Date인가요? Date이 어떤 날짜 형식을 말하는건가요?

Ryan-Dia replied:
감사합니다. ☺️
해당 부분 수정하였습니다.

kang-hyungu commented on src/main/java/domain/AttendanceCounter.java:
```diff
@@ -0,0 +1,31 @@
+    private final int absenceCount;
+
+    public static AttendanceCounter of(final Attendances attendances) {
+        AttendanceDto dto = attendances.calculateAttendanceCount();
```
왜 여기서 dto를 쓰고 있나요?

Ryan-Dia replied:
알려주셔서 감사합니다.!
저희도 다시보면서 왜 여기서 dto를 썼는지 의문입니다.  

해당 부분은 현재 아래처럼 수정되었습니다.
```java
  public static AttendanceCounter of(final Attendances attendances) {
        attendanceStatusCounts counts = attendances.calculateAttendanceCount();
        return new AttendanceCounter(counts.attendanceCount(), counts.tardinessCount(), counts.absenceCount());
    }
```

kang-hyungu commented on src/main/java/domain/AttendanceDto.java:
```diff
@@ -0,0 +1,4 @@
+package domain;
+
+public record AttendanceDto(int attendanceCount, int tardinessCount, int absenceCount) {
```
dto를 사용한 이유가 뭔가요?

kang-hyungu commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,38 @@
+        if (attendanceTime.isBefore(startTime) || attendanceTime.isAfter(endTime)) {
+            return ABSENCE;
+        }
+        if (attendanceTime.isBefore(day.getAttendanceTime().plusMinutes(5)) || attendanceTime.equals(
```
if절의 조건이 너무 복잡하네요.
attendanceTime.isBefore(day.getAttendanceTime().plusMinutes(5))
attendanceTime.equals(day.getAttendanceTime().plusMinutes(5))
위 두 조건을 각각 메서드로 추출하고 적절한 이름을 붙여보세요.

Ryan-Dia replied:
맞습니다. 아직 리팩터링을 하지 않아서 저도 복잡해보입니다. 😰 
그래서 아래처럼 개선했습니다.

```java
public static AttendanceStatus findByAttendanceTime(Week day, LocalTime attendanceTime) {
        if (isAbsence(attendanceTime)) {
            return ABSENCE;
        }
        if (isAttendance(day, attendanceTime)) {
            return ATTENDANCE;
        }
        if (isTardiness(day, attendanceTime)) {
            return TARDINESS;
        }
        return ABSENCE;
    }

    private static boolean isAbsence(final LocalTime attendanceTime) {
        return attendanceTime.isBefore(START_TIME) || attendanceTime.isAfter(END_TIME);
    }

    private static boolean isAttendance(final Week day, final LocalTime attendanceTime) {
        return isWithinTimeRange(attendanceTime, day.getAttendanceTime(), 5);
    }

    private static boolean isTardiness(final Week day, final LocalTime attendanceTime) {
        final LocalTime attendanceDeadline = day.getAttendanceTime().plusMinutes(5);
        return attendanceTime.isAfter(attendanceDeadline)
                && isWithinTimeRange(attendanceTime, day.getAttendanceTime(), 30);
    }

    private static boolean isWithinTimeRange(final LocalTime attendanceTime, final LocalTime baseTime,
                                             int rangeMinutes) {
        final LocalTime adjustedTime = baseTime.plusMinutes(rangeMinutes);
        return !attendanceTime.isAfter(adjustedTime);
    }
```

kang-hyungu commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,38 @@
+                day.getAttendanceTime().plusMinutes(5))) {
+            return ATTENDANCE;
+        }
+        if (attendanceTime.isBefore(day.getAttendanceTime().plusMinutes(30)) || attendanceTime.equals(
```
여기도 마찬가지로 적절한 메서드를 추출해보세요.

kang-hyungu commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,38 @@
+    TARDINESS("지각"),
+    ABSENCE("결석");
+
+    private final String koreanName;
```
왜 korean이라는 이름이 붙어야 하죠?

Ryan-Dia replied:
다시보니 타당한 이유가 없습니다.
따라서 koreanName에서 displayName으로 변경하였습니다. 😊

kang-hyungu commented on src/main/java/domain/Attendances.java:
```diff
@@ -0,0 +1,62 @@
+        this.attendances = attendances;
+    }
+
+    public AttendanceDto calculateAttendanceCount() {
```
dto를 쓰는 이유가 무엇일까요?

kang-hyungu commented on src/main/java/domain/Attendances.java:
```diff
@@ -0,0 +1,62 @@
+        final long tardinessCount = attendances.stream()
+                .filter(attendance -> attendance.attendanceStatus.equals(AttendanceStatus.TARDINESS))
+                .count();
+        final long absence = attendances.stream()
```
비슷한 패턴의 코드가 보이네요
중복을 제거할 방법이 없을까요?

asebn1 replied:
추가로 참고하면 좋을 자료 공유드립니다

- [스트림은 주의해서 사용하라](https://velog.io/@semi-cloud/Effective-Java-%EC%95%84%EC%9D%B4%ED%85%9C-45-%EC%8A%A4%ED%8A%B8%EB%A6%BC%EC%9D%80-%EC%A3%BC%EC%9D%98%ED%95%B4%EC%84%9C-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC)

Ryan-Dia replied:
맞습니다. 중복이 보기 좋지 않습니다. 
"일단 쓰세요 나중에 리팩터링 하면되요" 하고 시간없어서 못한게 후회되네요  
해당 부분은 아래처럼 수정하였습니다.
```java
 public attendanceStatusCounts calculateAttendanceCount() {
        Map<AttendanceStatus, Long> counts = attendances.stream()
                .collect(Collectors.groupingBy(
                        Attendance::getAttendanceStatus,
                        Collectors.counting()
                ));

        final long attendanceCount = counts.getOrDefault(AttendanceStatus.ATTENDANCE, 0L);
        final long tardinessCount = counts.getOrDefault(AttendanceStatus.TARDINESS, 0L);
        final long absenceCount = counts.getOrDefault(AttendanceStatus.ABSENCE, 0L);

        return new attendanceStatusCounts((int) attendanceCount, (int) tardinessCount, (int) absenceCount);
    }
```

참고자료 첨부해주셔서 진심으로 감사드립니다.

kang-hyungu commented on src/main/java/domain/Attendances.java:
```diff
@@ -0,0 +1,62 @@
+
+    public Attendance findAttendance(final LocalDateTime dateTime) {
+        return attendances.stream()
+                .filter(attendance -> attendance.localDateTime.toLocalDate().equals(dateTime.toLocalDate()))
```
1. attendance 객체 내부에서 처리하도록 메서드를 추가해보면 어떨까요?
2. Attendance 객체의 localDateTime 필드를 Attendances 객체가 접근하도록 허용해도 문제 없나요?

Ryan-Dia replied:
감사합니다. 
1. getter보다 메시지를 보내서 해당 객체에서 일을 하도록 변경하였습니다.
```java
    public Attendance findAttendance(final LocalDateTime dateTime) {
        return attendances.stream()
                .filter(attendance -> attendance.hasSameDate(dateTime.toLocalDate()))
                .findFirst()
                .orElseThrow(() -> new CustomIllegalArgumentException("수정하는 일자를 찾을 수 없습니다."));
    }
```

2. 무조건 문제가 있습니다.   
접근제어자를 사용하지 않고 필드를 외부에 공개하면, 객체의 내부 상태가 외부에 노출되어 캡슐화 원칙이 무너집니다. 

특히 외부에서 객체의 필드를 직접 수정할 수 있으므로, 잘못된 값이나 불법적인 값이 들어가 객체의 상태가 엉망이 될 수 있습니다.
또한 내부 구현을 변경할 때, 해당 필드를 직접 참조하고 있는 모든 코드를 함께 수정해야 하므로 코드의 유지보수가 복잡해집니다.
끝으로 내부 데이터가 공개되면, 이를 악용하는 경우가 발생할 수 있어 보안에 취약해집니다.


사실 제가 제일 놀랐습니다.  접근제어자와 fianl 필드가 어디갔지 했습니다. 😱

kang-hyungu commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,65 @@
+
+public class Crew implements Comparable<Crew> {
+
+    Nickname nickname;
```
접근제어자를 생략한 이유가 무엇인가요?

Ryan-Dia replied:
TDD할 때 필드까지만 만들어 놓고 하는데 이때 놓쳤던 것 같습니다. 
저도 놀랬습니다.

kang-hyungu commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,65 @@
+    Attendances attendances;
+    AttendanceCounter attendanceCounter;
+
+    public Crew(final Nickname nickname, Attendances attendances, AttendanceCounter attendanceCounter) {
```
final을 붙이려면 일관되게 붙여주세요.

Ryan-Dia replied:
네 현재 final이 필요한 모든 곳에 final을 추가했습니다.
다음에는 처음 구현 단계부터 나중에 달아야겠다고 생각하지않고 바로 작성하겠습니다.

kang-hyungu commented on src/main/java/domain/Crews.java:
```diff
@@ -0,0 +1,51 @@
+            .comparingInt((Crew crew) -> Punishment.findByAbsenceCount(getPunishment(crew)).getAbsenceCount())
+            .reversed();
+
+    private static int getPunishment(final Crew crew) {
```
메서드명에 get을 사용한 이유가 뭔가요?

Ryan-Dia replied:
`get`은 보통 단순히 객체의 필드 값을 반환할 때 사용하는 이름인데 해당 부분에서는 잘못사용해서 혼동을 드렸던 것 같아서 죄송합니다. 
해당 부분은 find로 변경하였습니다.

kang-hyungu commented on src/main/java/domain/Crews.java:
```diff
@@ -0,0 +1,51 @@
+    public List<Crew> getSortedCrews() {
+        return crews.stream().
+                sorted(
+                        punishmentOrder
```
정렬 조건에 줄바꿈이 너무 많은데요
차라리 정렬 조건 메서드로 추출하는건 어떤가요

Ryan-Dia replied:
네 저도 메서드로 추출하는 것이 가독성 측명해서 월등히 더 좋다고 생각하여 아래처럼 수정해보았습니다.

```java
private List<Crew> getSortedCrews() {
        return crews.stream().
                sorted(sortCrew())
                .toList();
    }

    private Comparator<Crew> sortCrew() {
        return punishmentOrder.thenComparing(absenceOrder)
                .thenComparing(tardinessOrder)
                .thenComparing(nameOrder);
    }

```

kang-hyungu commented on src/main/java/domain/Crews.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    private final Comparator<Crew> punishmentOrder = Comparator
+            .comparingInt((Crew crew) -> Punishment.findByAbsenceCount(getPunishment(crew)).getAbsenceCount())
```
규칙 4: 한 줄에 점을 하나만 찍는다.

Ryan-Dia replied:
감사합니다. 해당 부분 아래처럼 수정하였습니다.
```java
    private final Comparator<Crew> punishmentOrder = Comparator
            .comparingInt(this::getCrewPunishmentAbsenceCount)
            .reversed();
```

kang-hyungu commented on src/main/java/domain/Crews.java:
```diff
@@ -0,0 +1,51 @@
+
+    private static int getPunishment(final Crew crew) {
+        final AttendanceCounter attendanceCounter = crew.getAttendanceCounter();
+        return (attendanceCounter.getTardiness() * 3) + attendanceCounter.getAbsence();
```
3이 뭔가요?

Ryan-Dia replied:
매직넘버를 놓친 것 같습니다. 
해당부분은 지각 3회는 결석 1회로 간주할 때 필요한 3입니다. 
그래서 아래처럼 상수화를 진행했습니다.

```java
private static final int TARDINESS_TO_ABSENCE_RATIO = 3;
```

kang-hyungu commented on src/main/java/domain/Nickname.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private void validateNickname(final String nickname) {
+        String regex = "^[가-힣]{2,4}$";
```
검증에서 정규표현식을 사용해야 할까요?
조건이 복잡해지면 정규표현식도 복잡해질텐데, 그럼 어떤 조건이 있는지 파악하기 점점 어려워지지 않을까요?

Ryan-Dia replied:
네 저는 오히려 간단할때는 정규표현식을 적극적으로 활용합니다.  
해당 부분보다 더 복잡해지면 그때 변경해도 늦지 않을 것 같습니다.  
하지만 제가 속한 팀에서 정규표현식을 지양한다면 사용하지 않을 것 같습니다. 

페어랑 이야기가 되었기 때문에 이번 프로젝트에서는 의사소통에 문제가 없어서 사용했습니다.  
아직은 멀리보는 힘이 부족합니다.   
그렇기에 이렇게 사용해보고 추후 프로젝트가 커진다면 "아 그때 코치님이 말씀하신게 이거구나"를 직접 느껴보겠습니다.

kang-hyungu commented on src/main/java/domain/Punishment.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+
+    public static Punishment findByAbsenceCount(int absenceCount) {
```
왜 여기서는 final을 안 붙였나요?

kang-hyungu commented on src/main/java/domain/Week.java:
```diff
@@ -0,0 +1,58 @@
+
+public enum Week {
+
+    MONDAY(LocalTime.of(13, 0)),
```
여기는 왜 13인가요?

kang-hyungu commented on src/main/java/domain/Week.java:
```diff
@@ -0,0 +1,58 @@
+
+    private final LocalTime attendanceTime;
+
+    private static Map<DayOfWeek, String> initWeekNames() {
```
메서드 순서에서 생성자를 먼저 써주세요.

kang-hyungu commented on src/main/java/domain/Week.java:
```diff
@@ -0,0 +1,58 @@
+                .findFirst()
+                .orElseThrow(() -> new CustomIllegalArgumentException(
+                        String.format("%d월 %d일 %s은 등교일이 아닙니다.", Constants.MONTH,
+                                localDateTime.getDayOfMonth(), dayOfWeek)));
```
한줄에 너무 길게 쓰면서 줄바꿈을 하기보다, 문자열 부분을 변수로 추출해보세요.
줄바꿈도 의미를 담아서 하면 코드 파악에 도움이 됩니다.

Ryan-Dia replied:
맞습니다.  해당 부분은 제가봐도 한 눈에 코드를 파악하기 어려운 것 같습니다. 
말씀해주신 내용을 반영해서 아래처럼 변경해 보았습니다.

```java
public static Week findByAttendanceTime(final LocalDateTime localDateTime) {
        final String format = String.format(localDateTime.format(NON_SCHOOL_DAY_FORMAT) + "은 등교일이 아닙니다.");
        final DayOfWeek day = localDateTime.getDayOfWeek();

        return Arrays.stream(Week.values())
                .filter(week -> week.name().equals(day.name()))
                .findFirst()
                .orElseThrow(() -> new CustomIllegalArgumentException(format));
    }
```

kang-hyungu commented on src/main/java/util/Constants.java:
```diff
@@ -0,0 +1,16 @@
+import java.time.LocalTime;
+import java.util.List;
+
+public final class Constants {
```
상수를 하나의 클래스에서 관리하면 전역변수처럼 변질될 우려가 있습니다.
어디서 쓰는지 찾아보고 관리하기도 힘들어요.
상수가 필요한 클래스에 적절히 분배하세요.

kang-hyungu commented on src/main/java/util/CrewGenerator.java:
```diff
@@ -0,0 +1,91 @@
+    public static Crews generate(List<String[]> parsedCrewsData, LocalDate nowDate) {
+        Map<Nickname, List<Attendance>> crewData = new HashMap<>();
+        for (String[] parsedCrewData : parsedCrewsData) {
+            String nickname = parsedCrewData[0];
```
0, 1이 무엇을 의미하나요?

kang-hyungu commented on src/main/java/util/CrewGenerator.java:
```diff
@@ -0,0 +1,91 @@
+    private CrewGenerator() {
+    }
+
+    public static Crews generate(List<String[]> parsedCrewsData, LocalDate nowDate) {
```
메서드가 10줄이 넘지 않도록 수정해보세요

Ryan-Dia replied:
네 ! 
해당 부분은 리팩터링을 못해서 코드 보시기에 많이 불편하셨을 것 같습니다.  

아래처럼 리팩터링 해보았습니다.

```java
public static Crews generate(final List<String[]> parsedCrewsData, final LocalDate nowDate) {
        final Map<Nickname, TreeSet<Attendance>> crewData = generateCrewData(parsedCrewsData);
        final List<Integer> validDates = getValidDates(nowDate);
        final List<Crew> crews = new ArrayList<>();

        processCrewData(crewData, validDates, crews);

        return new Crews(crews);
    }
```

kang-hyungu reviewed (changes requested):
새로이~ 몇 가지 피드백 남겼어요
리뷰가 길어질 것 같아서 한 번 끊고 갈게요.
코드 수정해보시고, 궁금한 점 생기면 언제든 질문주세요

Ryan-Dia committed:
style: 생성자 위치 변경

Ryan-Dia committed:
refactor: 매개변수에 final 키워드 추가

Ryan-Dia committed:
feat: AttendanceDateTime 객체 추가

- localDateTime이 쓰이는 곳이 많아 그냥 가지고 있지 않고 객체로 랩핑함

Ryan-Dia committed:
refactor: 정적 팩터리 메서드를 이용해서 결석 기록 로직 단순화

asebn1 commented on README.md:
```diff
@@ -1,3 +1,67 @@
 # java-attendance
 
+## 기능 목록
+
 출석 미션 저장소
+
+- 시간은 24시간 형식
+- 교육시간
+    - 월요일 13:00 ~ 18:00
+    - 화요일 ~ 금요일 10:00 ~ 18:00
+- 출결
+    - 지각 : 5분초과
+    - 결석
+        - 30분초과 or 출석기록 없을 때
+        - 지각 3회는 결석 1회로 간주
+    - 경고 대상자 : 결석 3회 초과
+    - 면담 대상자 : 결석 4회 초과
+    - 제적 대상자 : 결석 5회 초과
+- 운영 시간
+    - 08:00 ~ 23:00
+- 주말 및 공휴일에는 출석을 받지 않음
+- csv파일에서 확인할 수 있다.
+- 사용자가 종료할 때까지 종료되지 않음
+- 해당 기능을 수행한 후 초기 화면으로 돌아감
```
기능 목록을 잘 작성해주셨네요 👍 
다음에는 기능별 요구사항을`체크리스트`로 관리해보는 것도 제안드립니다 :)

Ryan-Dia replied:
네 ! 2단계에는 체크리스트로 활용해서 진행해보겠습니다.
알려주셔서 감사합니다.

asebn1 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,38 @@
+    private final String koreanName;
+
+    private static final LocalTime startTime = LocalTime.of(8, 0);
+    private static final LocalTime endTime = LocalTime.of(23, 0);
```
상수는 일반적으로 `대문자 + 언더스코어` 형식을 사용합니다

Ryan-Dia replied:
제가봐도 당황스럽습니다.😰
알려주셔서 감사합니다. !

asebn1 commented on src/main/java/util/CrewGenerator.java:
```diff
@@ -0,0 +1,91 @@
+        List<Integer> noPresentAttendanceDates = new ArrayList<>(validDates);
+        noPresentAttendanceDates.removeAll(alreadyAttendanceDates);
+        for (Integer attendanceDate : noPresentAttendanceDates) {
+            LocalDateTime dateTime = LocalDateTime.of(LocalDate.of(2024, Constants.MONTH, attendanceDate),
```
`2024`도 상수화해볼까요?

Ryan-Dia replied:
네 해당부분도 상수화 진행했습니다.

asebn1 commented on src/main/java/util/CrewGenerator.java:
```diff
@@ -0,0 +1,91 @@
+        return new Crews(crews);
+    }
+
+    private static Attendances getAttendances(final Entry<Nickname, List<Attendance>> nicknameListEntry,
```
대부분 메서드가 static으로 사용하고 있네요. 
이유가 궁금합니다 :)

Ryan-Dia replied:
저희는 상태값을 가지지 않고 인스턴스화를 할 필요가 없다고 판단하였고
객체의 인스턴스 변수나 상태와 상관없이 동작하는 메서드들의 모임이기에 static으로 해야겠다 생각했는데 

다시 생각해보니 일부메서드가 테스트하기 어렵다고 판단해서 다시 static을 제거하고 인스턴스화가 가능하도록 변경해서 필요한 변수를 주입받도록 변경했습니다.

asebn1 commented on src/test/java/domain/CommandTest.java:
```diff
@@ -0,0 +1,31 @@
+        );
+    }
+
+}
```
```suggestion
class CommandTest {

    @ParameterizedTest
    @MethodSource("methodSources")
    void 올바른_명령_문자를_입력시_명령_반환(String commandName, Command expectedCommand) {
        // given

        Command command = Command.findByCommandNumber(commandName);
        Assertions.assertThat(command).isEqualTo(expectedCommand);
    }

    private static Stream<Arguments> methodSources() {
        return Stream.of(
                Arguments.arguments("1", Command.CHECK_ATTENDEES),
                Arguments.arguments("2", Command.EDIT_ATTENDANCE),
                Arguments.arguments("3", Command.CHECK_THE_ATTENDANCE_RECORD_BY_CREW),
                Arguments.arguments("4", Command.CONFIRMATION_OF_THOSE_AT_RISK_OF_EXPULSION),
                Arguments.arguments("Q", Command.QUIT)
        );
    }
}
```
의미 없는 개행은 제거해주세요

Ryan-Dia replied:
감사합니다. 다음에는 조금 더 신경쓰도록 하겠습니다.!

asebn1 commented on src/test/java/domain/NicknameTest.java:
```diff
@@ -0,0 +1,34 @@
+
+class NicknameTest {
+
+    @Test
+    void 올바른_닉네임을_입력한다면_통과() {
+        // given
+        String input = "칼리";
+
+        // when
+        final Nickname nickname = new Nickname(input);
+
+        // then
+        assertThat(nickname.getNickname()).isEqualTo(input);
+    }
```
깔끔하게 테스트를 잘 작성해주셨네요 💯

asebn1 commented on src/main/java/util/CrewGenerator.java:
```diff
@@ -0,0 +1,91 @@
+    public static List<Integer> getValidDates(LocalDate localDate) {
+        final int today = localDate.getDayOfMonth();
+
+        List<Integer> allDays = IntStream.range(1, today).boxed().collect(Collectors.toList());
```
객체지향 생활체조 원칙을 지켜볼까요?

- [참고](https://tmd8633.tistory.com/12)

Ryan-Dia replied:
네 해당 부분 참고해서 수정했습니다. 감사합니다.!

asebn1 reviewed (changes requested):
새로이 안녕하세요 :)
이번 출선구현미션 리뷰어 제로입니다.
제 과거 유산물(?)을 보셨다니..!! 살짝 부끄러워지네요
저도 과거에 경험한 만큼 해주신 이야기들이 많이 공감됩니다.
구구가 먼저 도움되는 피드백을 대부분 많이해주셔서 저는 추가로만 코멘트 드렸습니다.
이번 미션도 화이팅입니다 🔥🔥

---

> 1. InputView 메서드 내에서 String 반환 (의존성 제거)

결론 먼저 말씀드리면 현 코드 로직이 좋다고 생각합니다. View는 도메인 모델에 대한 의존을 갖지 않는 것이 좋다고 생각해요. 그 이유는 무엇일까요-?

> 2. 같은 로직을 처리하는 메서드 관리

지금 로직에서 매개변수 활용하는 로직이 코드 개선이 될 것 같아요. 사용할 땐 아래처럼 사용할 수 있을 것 같아요.
```
String nickname = readInput("닉네임을 입력해 주세요.");
String updateNickname = readInput("출석을 수정하려는 크루의 닉네임을 입력해 주세요.");
...
```

> 3. 정렬 위치 view?

저도 같은 의견입니다! 
다만 도메인에서 애초에 항상 정렬된 상태를 유지해야한다면 도메인이 더 적절할 것 같고, 그게 아니라면 어떤 곳에 위치하는게 더 적절한 책임을 갖는지 고민하고 구현할 것 같아요.

> 4. 출석 가능한 일자 뽑아오는 로직의 효율성

말씀 주신 아이디어(1. 첫 번째 토요일과 일요일을 찾기, 2. 각 요일에 +7씩 더하면서 리스트에 추가하기)도 좋다고 생각합니다. 저도 비슷한 방법으로 구현하지 않을까 싶어요.

> 5. class 이름과 같은 필드 이름

좋은 고민입니다. 예약어까지 생각해주셨군요 👍 
예약어는 피하는 것은 좋다고 생각합니다.
네이밍은 실제 실무에서도 많이 고민되는 포인트인데요, 주변 동료분들과 함께 고민하기도합니다.
name은 보통 `username`으로 사용하고, 다른 네이밍들은 적절하게 결정할 것 같아요.

6. Test 코드, DisplayName 애노테이션 사용 유무

`@DisplayName`를 활용하는 것이 좋다고 생각하고 이것도 명확한 정답은 아니라고 생각해요.
팀 테스트 컨밴션에 맞추는 것이 좋다고 생각합니다 :)

7. 출력 시, StringFormat vs “\n” 매번 작성

`System.out.println`를 직접 사용하는 경우가 많이 없는데요, 지금은 과제가 출력이 필요한 요구사항이기에 궁금하신 것 같아요 ㅎㅎ 보통 로그나 템플릿 기반 문자열 처리 방식을 주로 사용한다고 생각해주시면 좋을 것 같아요.

> 8. 의존성 depth 때문에 같은 이름의 메서드 계속 생성 
저의 생각은 Nickname 클래스에서 nickname 필드 이름을 displayNickname으로 바꿔서 Crew 객체에서 getDisplayNickname() 이란 메서드를 만들어서 가져올려고 했습니다.

좋은 해결 방법이라고 생각합니다 💯

> 9. 명령어 입력 controller if 처리

switch-case로 한번 개선포인트를 잡아보면 어떨까요?

---

우선 질문드린 내용에대해서 코멘트 드립니다
명확한 답을 드리는 것보다 직접 한번 찾아보면 학습에 도움이 되실 것 같아 이렇게 답변드려요

Ryan-Dia committed:
feat: 반복 루프 유틸 추가

최대 시도 횟수는 100회로 제한함

Ryan-Dia committed:
refactor: controller에 커맨드 타입 도입

명령이 추가될 때마다 if문이 계속 증가하는 문제 발생

Ryan-Dia committed:
feat: 출석 시간 도메인 클래스 추가

- 문자열을 LocalTime으로 파싱하는 AttendanceTime 클래스 구현
- 올바르지 않은 시간 포맷 시 CustomIllegalArgumentException 발생 처리

Ryan-Dia committed:
feat: 오늘의 LocalDateTime을 가진 AttendanceDateTime을 반환하는 정적 팩터리 메서드 추가

Ryan-Dia committed:
refactor: 접근 제어자를 public으로 변경

다른곳에서도 사용하기에 public으로 변경

Ryan-Dia committed:
refactor: readLocalTime 메서드 단순화

기존에 시간을 파싱하는 책임을 AttendanceTime으로 이관함

Ryan-Dia committed:
refactor: 제적 위험자 확인 개선

- 결석 보정 계산을 AttendanceCounter 객체가 하도로 변경
- crews를 바로 view에 넘겨주지 않고 필요한 데이터만 뽑아서 넘겨주도록 변경
- crew 필드에 접근 제어자 및 final 추가
- crew 필드를 생성하는 메서드 추가

Ryan-Dia committed:
refactor: attendance 컬렉션 변경 list-> treeSet

자동정렬을 위해 list를 treeSet으로 변경

Ryan-Dia committed:
refactor: 기존 한글화 로직을 한글 포맷팅으로 대체

Ryan-Dia committed:
feat: 크루 서머리에 출석 횟수 추가

Ryan-Dia committed:
feat: 출석 서머리 추가

attendanceDateTime과 attendanceStatus를 한 번에 가져옴

Ryan-Dia committed:
refactor: 날짜 한글화 기능을 포메팅으로 변경

Ryan-Dia committed:
refactor: final 키워드 및 접근제어자 private으로 변경

Ryan-Dia committed:
refactor: 가독성을 위해 메서드 분리

Ryan-Dia committed:
refactor: 필드에 바로 접근이아닌 getter을 통해 접근하도록 변경

Ryan-Dia committed:
style: 클래스명 변경 (AttendanceDto -> AttendanceStatusCounts)

Ryan-Dia committed:
refactor: stream을 한 번만 사용하도록 변경

Ryan-Dia committed:
style: 첫째줄을 스킵하는 이유 주석으로 추가

Ryan-Dia committed:
refactor: 정렬 로직 메서드로 분리

Ryan-Dia committed:
refactor: 가독성을 위해 format을 변수로 분리

Ryan-Dia committed:
style: 가독성을 위해 코드 개행

Ryan-Dia committed:
refactor: 상수모음 객체 삭제

Ryan-Dia committed:
style: 상수명 변경

FIXED -> REFFERENCE
FIXED_DATE_TIME->SYSTEM_DATE_TIME

Ryan-Dia committed:
style: 예외 메시지 수정

Ryan-Dia committed:
refactor: 가독성을 위해 조건문 메서드로 추출

Ryan-Dia committed:
style: 변수명 변경 (koreanName->displayName)

Ryan-Dia committed:
refactor: getter가아닌 메시지를 보내도록 로직 변경

Ryan-Dia committed:
refactor: 메서드 분리

Ryan-Dia committed:
style: 의미없는 개행 제거

Ryan-Dia committed:
style: stream로직에 개행 추가

Ryan-Dia committed:
refactor: 사용하지 않는 변수 삭제

Ryan-Dia committed:
refactor: getter 대신 메시지를 보내서 결과값을 가져오도록 수정

Ryan-Dia committed:
refactor: 로직 메서드로 분리

Ryan-Dia committed:
style: 클래스명 첫글자 대문자로 수정

Ryan-Dia committed:
style: 클래스명 첫 글자 대문자로 수정

Ryan-Dia committed:
style: command 인터페이스 디렉토리 이동

Ryan-Dia committed:
refactor: 클래스 첫 글자 대문자로 수정

Ryan-Dia committed:
style: 잘못된 메서드명 수정

Ryan-Dia committed:
refactor: 줄바꿈자를 System.lineSeparator()로 수정

Ryan-Dia committed:
feat: 출석 확인 기능에서 등교일이 아니면 선제적으로 예외를 하도록 변경

Ryan-Dia committed:
refactor: CrewGenerator를 인스턴스화가 가능하도록 수정

Ryan-Dia committed:
fix: 출석시간을 매개변수와 필드를 비교하도록 수정

Ryan-Dia committed:
refactor: 정적 팩터리 메서드 네이밍 변환

Ryan-Dia commented:
# 🛠️주요 개선 사항

## 1️⃣ 인터페이스를 활용해서 명령마다 로직 분리


<img width="628" alt="image" src="https://github.com/user-attachments/assets/ca9373c2-54a6-4f21-b8ff-b8fa2707ac6e" />

예를들어 정말 나중에 명령이 20개가 넘어간다면 20개의 if절이 필요하고 그럴수록  controller는 엄청나게 비대해질 것 같습니다.  따라서 다른 방법으로 구현해야할 필요를 느꼈습니다.  

그러던중 인터페이스를 활용해서 명령별로 나누는 방법 즉, 다형성을 이용해 명령마다 객체를 만드는 방법에 대해 알게되었습니다.  

```java
public void run() {
        final Crews crews = CrewGenerator.generate(CsvReader.readFile(CSV_PATH), SYSTEM_DATE_TIME.toLocalDate());

        Loop.run(() -> {
            final Command commandEnum = Command.findByCommandNumber(InputView.readCommand(SYSTEM_DATE_TIME));
            final AttendanceCommand command = commandEnum.getCommandInstance();

            if (command == null) {
                return false;
            }
            command.execute(crews);
            return true;
        });
    }
}
```

제가 생각하기에 위 방법을 사용하면 아래의 점들이 좋습니다. 

### 1. 단일 책임 원칙

if문이 없어지는 대신  명령마다 클래스를 생성해야하기 때문에 클래스는 많아지지만 각각의 클래스가 하나의 기능만 담당하게 됩니다.  

이전 처럼 controller에 20개의 if문이 돌아가기위해 여러 로직이 한 덩어리로 얽혀 있으면, 어떤 로직을 수정할 때 다른 로직도 건드릴 위험이 커집니다. 반면에 명령이 분리되어 있다면, 해당 기능만 딱 수정하면 되니까 안전합니다.

### 2. 개방 폐쇄 원칙

기존 코드를 거의 건드리지 않고 새로운 명령을 바로 추가할 수 있습니다.

if문이 20개가 있는 클래스에 새 기능을 추가하려면 기존 코드에 추가를 해주게되어 controller는 점점 더 비대해집니다.

반면 명령별로 클래스를 나누면 enum에 추가해주고 클래스만 만들어주면 되기 때문에 훨씬 간편합니다.

### 3. 가독성 증가 및 유지보수 비용 하락

if문이 20개가 있다면 위에서부터 쭉 읽으면서 파악을 해야합니다.  그리고 중간에 모르는 메서드가 있다면 왔다갔다 해야합니다.  하지만 명령마다 클래스가 분리되어있다면  클래스명만 보고 바로 찾아서 볼 수 있으며  해당 클래스에는 해당 명령에 관한 메서드만 있기 때문에 훨씬 가독성이 높습니다.   

그리고 이 가독성을 통해 유지보수가 훨씬 수월해집니다.

**📚 정리**

if문 20개  vs 클래스 20개를 본다면 표면적으로 보면 둘 다 많다고 느낄 수 있지만  클래스 분리는 소스가 명확하게 나뉘고, 변경/추가가 용이하다는 큰 이점이 있습니다. 그리고 프로젝트가 커질수록 하나의 파일에 if문이 주르륵 있는 것보다는 각 기능이 작은 단위로 독립되어있는 구조가 훨씬 유지보수하기에 좋습니다.  

결론적으로 클래스가 많아지는 불편보다 유연성, 가독성, 유지보수성 향상이 훨씬 큰 이득가져온다고 생각하기에 이렇게 바꾸었습니다. 

코드를 한 번 작성하면 끝이아니라 계속 수정, 확장, 유지보수를 해야하기에 이러한 점들이 더 좋은 것 같습니다. 

## 2️⃣  왜 CrewGenerator에는 모든 메서드가 static인가?

CrewGenerator 클래스는 단순히 데이터를 가공하고 결과를 생성하는, 상태를 갖지 않는 순수한 기능들을 제공하기 때문에 모든 메서드를 static으로 작성하였습니다. 

하지만 문제점을 발견했습니다. 

현재는 AttendanceCommandController의 상수처럼 외부에 고정된 값을 일부 메서드가 사용하기 때문에 국소적으로만 테스트가 가능합니다. 

인스턴스화하면 REFERENCE_YEAR, REFERENCE_MONTH 같은 설정 정보를 생성자를 통해 주입받아 다양한 환경에 맞게 테스트가 가능할 것 같습니다. 

## 3️⃣  최대 시도 횟수 제한

무한대로 시도할 수 있게하는 게 아니라 일정 횟수 이상은 비정상적인 동작으로 판별하여 로직을 종료합니다.  

현재 최대 시도 횟수 100회
 
## 4️⃣ 출석 시간을  객체로 래핑

기존에는 localDateTime을 Attendance내부에서 검증 한 뒤에 사용했는데 생각 보다 검증이 복잡하고 사용하는 곳이 많아서 아예 객체로 래핑해서 정말 출석하는 데 유효한 시간만 갖는 것을 보장하자고 생각했습니다. 

래핑한 결과 이전에는 localDateTime 무엇을 의미하는지 저만 알았는데 이제는 출석과 관련된 시간임이 명확해졌습니다. 또한 출석 시간에 대한 비즈니스 로직을 AttendanceDateTIme에서 수행할 수 있어서 Attendance가 훨씬 가벼워졌습니다.

**🔴 Before**

```java
public class Attendance implements Comparable<Attendance> {

    private final LocalDateTime localDateTime;
    private final AttendanceStatus attendanceStatus;
}
```

**✅ After**

```java
public class Attendance implements Comparable<Attendance> {

    private final AttendanceDateTime attendanceDateTime;
    private final AttendanceStatus attendanceStatus;
}
```

## 5️⃣ getter로 꺼내는 대신 메시지를 보내도록 수정

일부로직에서 무의식적으로 getter을 사용해서 비교하고있 던 코드를 수정하였습니다.
**🔴 Before**

```java
attendances.stream()
            .filter(attendance -> attendance.getLocalDateTime().toLocalDate().equals(dateTime.toLocalDate()))
```

**✅ After**

```java
attendances.stream()
            .filter(attendance -> attendance.hasSameDate(dateTime.toLocalDate()))
```

## 6️⃣ 정적 팩터리 메서드를 사용할 때는  생성자를 private로 변경

기존에는 정적 팩터리 메서드를 사용하더라도 외부에서 직접 인스턴스를 생성할 수 있도록 생성자를 public으로 열어두었습니다.
하지만 이번 리팩터링 과정에서 이에 대해 다시 고민해 보았습니다.

생성자가 public이면 누구나 new 키워드를 사용하여 인스턴스를 생성할 수 있으므로, 정적 팩터리 메서드가 제공하는 제어 및 유연성의 장점을 충분히 활용하기 어렵다고 판단했습니다.

또한, 생성자를 private으로 지정하면 상속이 제한되어 컴포지션을 사용하도록 유도할 수 있고, 불변 타입을 유지하는 데에도 도움이 됩니다.

(상속을 통해 불변성을 깨뜨릴 가능성이 생길 수 있습니다.)

다만, 상속을 허용하거나 리플렉션을 사용하는 프레임워크와의 호환성을 고려해야 하는 경우에는 생성자를 public으로 해둘 것 같습니다. 

```java
private Attendance(final AttendanceDateTime attendanceDateTime) {
        final Week day = Week.findByAttendanceTime(attendanceDateTime.getLocalDateTime());
        this.attendanceDateTime = attendanceDateTime;
        this.attendanceStatus = AttendanceStatus.findByAttendanceTime(day, attendanceDateTime.toLocalTime());
  }

public static Attendance getInstance(AttendanceDateTime attendanceDateTime) {
    return new Attendance(attendanceDateTime);
}

public static Attendance from(final String inputTime) {
    return new Attendance(AttendanceDateTime.of(inputTime));
}
```


## 🤔 궁금한 점들

1. 팀바이팀이겠지만 리뷰어님의 생각이 궁긍합니다.
  - 1) 정적 팩터리 메서드를 쓰면 생성자는 private으로 거의 해두시는지?
  - 2) 정적 팩터리 메서드를 일부 클래스에서만 쓰면 다른 개발자들은 생성자를 바로 써야하는지 정적 팩터리메서드를 써야하는지 헷갈릴 수 있습니다.  그래서 현재 일하시고 계신곳에서는 어떻게 하시는 지 궁금합니다.

Ryan-Dia committed:
refactor: 방어적 복사가 되도록 변경

Ryan-Dia committed:
style: 의미 없는 개행 제거

Ryan-Dia committed:
refactor: crewDto에 담아 view에 전달되도록 로직 수정

Ryan-Dia committed:
refactor: dto를 컨트롤러에서 만들어서 주입하도록 변경

asebn1 commented on src/main/java/controller/command/AttendanceCommand.java:
```diff
@@ -0,0 +1,7 @@
+package controller.command;
+
+import domain.Crews;
+
+public interface AttendanceCommand {
+    void execute(Crews crews);
+}
```
`command` 패키지를 controller 패키지 내부에 정의한 이유가 있을까요?

asebn1 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,56 @@
+        return attendanceTime.isBefore(START_TIME) || attendanceTime.isAfter(END_TIME);
+    }
+
+    private static boolean isAttendance(final Week day, final LocalTime attendanceTime) {
+        return isWithinTimeRange(attendanceTime, day.getAttendanceTime(), 5);
+    }
+
+    private static boolean isTardiness(final Week day, final LocalTime attendanceTime) {
+        final LocalTime attendanceDeadline = day.getAttendanceTime().plusMinutes(5);
+        return attendanceTime.isAfter(attendanceDeadline)
+                && isWithinTimeRange(attendanceTime, day.getAttendanceTime(), 30);
+    }
```
5와 30의 의미도 정확하게 알 수 없는 것 같아요

asebn1 commented on src/main/java/domain/Crews.java:
```diff
@@ -0,0 +1,55 @@
+                .orElseThrow(() -> new CustomIllegalArgumentException("크루가 존재하지 않습니다."));
+    }
+
+    public List<Crew> getSortedCrews() {
+        return crews.stream().
+                sorted(sortCrew())
+                .toList();
+    }
```
```suggestion
    public List<Crew> getSortedCrews() {
        return crews.stream()
                .sorted(sortCrew())
                .toList();
    }
```

asebn1 commented on src/main/java/domain/Crews.java:
```diff
@@ -0,0 +1,55 @@
+
+    public Crew findByNickname(final Nickname nickname) {
+        return crews.stream()
+                .filter(crew -> crew.getNickname().equals(nickname))
```
값을 가져와서 검사하지 않고 객체에 물어보는 형태로 바꾸어볼까요?

- https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/

asebn1 reviewed (approved):
많은 고민과 코드개선을 위해 노력하신 것이 보이네요 🔥🔥
간단한 코멘트 남겼습니다 
2단계 미션에 추가로 반영해주세요 :)
이번 단계는 이만 머지하겠습니다.

---
>   정적 팩터리 메서드를 쓰면 생성자는 private으로 거의 해두시는지?

생성자를 private으로 제한하려고 노력하는 편입니다 :)


> 정적 팩터리 메서드를 일부 클래스에서만 쓰면 다른 개발자들은 생성자를 바로 써야하는지 정적 팩터리메서드를 써야하는지 헷갈릴 수 있습니다. 그래서 현재 일하시고 계신곳에서는 어떻게 하시는 지 궁금합니다.

팀 내 코드 스타일을 외부에 명확히 공개하기는 어렵지만, 팀원 수가 많은 프로젝트에서는 컨벤션이 다소 일관되지 않을 때도 있습니다. 기본적으로 프로젝트별 컨벤션을 맞추는 것을 우선으로 하고, 그다음으로는 정적 팩토리 메서드를 적절한 곳에 적용하려고 노력하고 있습니다.
